### PR TITLE
Add snapshot for prefix iterate.

### PR DESCRIPTION
**Pull request by SuperYoko**
#### What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### What problem(s) does this PR solve?
Issue(s) number: 

Description:
Use prefix to iter data on the same snapshot of rocksdb.

#### Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Comment by critical27** [comment]
Close https://github.com/vesoft-inc/nebula/issues/3710

**Comment by critical27** [comment]
Very quick fix 👍

File: src/kvstore/RocksEngine.h [hunk 1]
```diff
@@ -166,16 +174,19 @@ class RocksEngine : public KVEngine {
                                 std::unique_ptr<KVIterator>* iter) override;
 
   nebula::cpp2::ErrorCode prefix(const std::string& prefix,
-                                 std::unique_ptr<KVIterator>* iter) override;
+                                 std::unique_ptr<KVIterator>* iter,
```
Comment (darionyaphet) [inline src/kvstore/RocksEngine.h:177]: why don't use KVIterator* directly?

File: src/kvstore/RocksEngine.h [hunk 2]
```diff
@@ -146,6 +146,14 @@ class RocksEngine : public KVEngine {
     return walPath_.c_str();
   }
 
+  const void* GetSnapshot() override {
+    return db_->GetSnapshot();
+  }
+
+  void ReleaseSnapshot(const void* snapshot) override {
+    db_->ReleaseSnapshot(reinterpret_cast<const rocksdb::Snapshot*>(snapshot));
```
Comment (darionyaphet) [inline src/kvstore/RocksEngine.h:154]: `ReleaseSnapshot`  would remove from file system?

Comment (SuperYoko) [reply src/kvstore/RocksEngine.h:154]: No, It's only a memory snapshot.

File: src/kvstore/RocksEngine.h (hunk 1 above)
Comment (SuperYoko) [reply src/kvstore/RocksEngine.h:177]: Yes, it is bad to pass pointer of smart pointer. I don't know why it is coded this way. Maybe I should try to refactor it. But it do affect too many files.  Don't want do this in this pr.

**Comment by liwenhui-soul** [comment]
what does void* mean? it doesn't have a type?

**Comment by SuperYoko** [comment]
> what does void* mean? it doesn't have a type?

I wan't to  create a wrapper like iterator first, but it didn't  help much understanding. 
Anyway, Rocksdb Snapshot pointer should pass to prefix function, and there should be a container to save pointer. 

Maybe a warpper of vector <boost::any> is okey, but it still pass a pointer. I don't have another clear idear. WDYT?

**Review by critical27 — APPROVED** [review]

File: src/kvstore/RocksEngine.h (hunk 2 above)
Comment (darionyaphet) [reply src/kvstore/RocksEngine.h:154]: AFAIK snapshot will create some hard link about the previous SST files why don't remove them?

Comment (SuperYoko) [reply src/kvstore/RocksEngine.h:154]: > AFAIK snapshot will create some hard link about the previous SST files why don't remove them?

https://github.com/facebook/rocksdb/wiki/Snapshot

> A snapshot captures a point-in-time view of the DB at the time it's created. Snapshots do not persist across DB restarts.

**Review by darionyaphet — APPROVED** [review]

**Comment by HarrisChu** [comment]
https://github.com/vesoft-inc/nebula/issues/3782

seems db_upgrader crashed by this

**Comment by panda-sheep** [comment]
Good job

**Pull request merged**